### PR TITLE
feat: Add buildlog package

### PR DESCRIPTION
**Pull request by kkokufud**
## Context
Output and parse logs from build container.

## Objective
- Read log file asynchronously

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: buildlog/buildlog_test.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	})
+
+}
+
+type compareableLog struct {
+	ctx    context.Context
+	file   string
+	writer io.Writer
+	cancel context.CancelFunc
+}
```
Comment (sugarnaoming) [inline buildlog/buildlog_test.go]: I think this struct should be removed because not use this struct.

File: buildlog/buildlog_test.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+		}
+
+		errChan := l.Run()
+		_ = errChan
+
+		time.Sleep(3 * time.Second)
+		l.Stop()
+
+		go write(t, tmpFile.Name(), testInputsNotWritten)
+
```
Comment (sugarnaoming) [inline buildlog/buildlog_test.go]: To confirm that writing is completed, I think that a sleep function is necessary.

File: buildlog/buildlog_test.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+		expected := []string{
+			"2020-01-28 16:56:49 +0900 JST: test 1\n",
+			"2020-01-28 16:57:02 +0900 JST: test 2\n",
+		}
+		assert.Equal(t, strings.Join(expected, ""), writer.String())
```
Comment (sugarnaoming) [inline buildlog/buildlog_test.go]: It's hard to understand the difference between `TestRun` and `TestStop`, so it's a good idea to add a comment.

File: buildlog/buildlog.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+			l.output(reader)
+		}
+		time.Sleep(readInterval)
+	}
+}
+
+func (l log) output(reader *bufio.Reader) error {
+	line, _, err := reader.ReadLine()
+	if err != nil {
+		return nil
```
Comment (sugarnaoming) [inline buildlog/buildlog.go]: Should return error

File: buildlog/buildlog.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	go l.run(errChan)
+	return errChan
+}
+
+func (l log) Stop() {
+	l.cancel()
+	return
+}
+
+func (l log) run(errChan chan error) {
```
Comment (sugarnaoming) [inline buildlog/buildlog.go]: `chan error` does not need.
I think it is ok to continue the build if the log parsing fails.

**Comment by sugarnaoming** [comment]
Is there no need for error case test?

File: buildlog/buildlog_test.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+		err = file.Close()
+		if err != nil {
+			tb.Fatal(err)
+		}
+	}
+}
+
+func TestRun(t *testing.T) {
+	t.Run("success", func(t *testing.T) {
+		tmpFile, err := ioutil.TempFile("", "")
```
Comment (kkisic) [inline buildlog/buildlog_test.go]: Is it better to clean up `tmpFile`?

File: buildlog/buildlog_test.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+			"2020-01-28 16:57:02 +0900 JST: test 2\n",
+		}
+		assert.Equal(t, strings.Join(expected, ""), writer.String())
+	})
+
+}
+
+func TestStop(t *testing.T) {
+	t.Run("success", func(t *testing.T) {
+		tmpFile, err := ioutil.TempFile("", "")
```
Comment (kkisic) [inline buildlog/buildlog_test.go]: Same as above.

File: buildlog/buildlog_test.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+type compareableLog struct {
+	ctx    context.Context
+	file   string
+	writer io.Writer
+	cancel context.CancelFunc
+}
+
+func TestNew(t *testing.T) {
+	t.Run("success", func(t *testing.T) {
+		tmpFile, err := ioutil.TempFile("", "")
```
Comment (kkisic) [inline buildlog/buildlog_test.go]: Same as above.

File: buildlog/buildlog_test.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+		parent, cancel := context.WithCancel(context.Background())
+		writer := bytes.NewBuffer(nil)
+		l := log{
+			file:   tmpFile,
+			writer: writer,
+			ctx:    parent,
+			cancel: cancel,
+		}
+
+		errChan := l.Run()
```
Comment (kkisic) [inline buildlog/buildlog_test.go]: Do we need `errChan`?
Because it is unused.

File: buildlog/buildlog.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+		case <-l.ctx.Done():
+			break
+		default:
+			l.output(reader)
+		}
+		time.Sleep(readInterval)
+	}
+}
+
+func (l log) output(reader *bufio.Reader) error {
```
Comment (tk3fftk) [inline buildlog/buildlog.go]: Do we use this `error`?

File: buildlog/buildlog_test.go (hunk 1 above)
Comment (kkokufud) [reply buildlog/buildlog_test.go]: Fixed.

File: buildlog/buildlog_test.go (hunk 2 above)
Comment (kkokufud) [reply buildlog/buildlog_test.go]: Fixed.

File: buildlog/buildlog_test.go (hunk 3 above)
Comment (kkokufud) [reply buildlog/buildlog_test.go]: I added description for TestStop.

File: buildlog/buildlog.go (hunk 4 above)
Comment (kkokufud) [reply buildlog/buildlog.go]: Fixed.

File: buildlog/buildlog_test.go (hunk 6 above)
Comment (kkokufud) [reply buildlog/buildlog_test.go]: Fixed.

File: buildlog/buildlog_test.go (hunk 7 above)
Comment (kkokufud) [reply buildlog/buildlog_test.go]: Fixed.

File: buildlog/buildlog_test.go (hunk 8 above)
Comment (kkokufud) [reply buildlog/buildlog_test.go]: Fixed.

File: buildlog/buildlog_test.go (hunk 9 above)
Comment (kkokufud) [reply buildlog/buildlog_test.go]: Fixed.

File: buildlog/buildlog.go (hunk 10 above)
Comment (kkisic) [reply buildlog/buildlog.go]: I fixed it.

File: buildlog/buildlog.go (hunk 5 above)
Comment (kkisic) [reply buildlog/buildlog.go]: I fixed it.

**Comment by sakka2** [comment]
Do we need the test case that the parse error happens?
This point.
https://github.com/screwdriver-cd/sd-local/pull/5/files#diff-ddceb8bc873c84a6fbc947fa42dcb3c4R97

File: buildlog/buildlog_test.go [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+		l := log{
+			file:   tmpFile,
+			writer: writer,
+			ctx:    parent,
+			cancel: cancel,
+		}
+
+		go l.Run()
+
+		time.Sleep(3 * time.Second)
```
Comment (tk3fftk) [inline buildlog/buildlog_test.go]: Do we need to wait 3 seconds?  I tried a second or 500 milliseconds and both is ok.
And this magic number appears 3 times so we should make this wait time a variable. 

note: any `sleep` cause running time of tests longer so we should avoid using it.

**Comment by kkisic** [comment]
> Do we need the test case that the parse error happens?

thx.
I added test case.

File: buildlog/buildlog.go [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+			}
+		}
+		time.Sleep(readInterval)
+	}
+}
+
+func (l log) output(reader *bufio.Reader) error {
+	line, _, err := reader.ReadLine()
+	if err != nil {
+		if err != io.EOF {
```
Comment (tk3fftk) [inline buildlog/buildlog.go]: Is `io.EOF` caused when "logfile is empty" or "all log line are read" ?

File: buildlog/buildlog_test.go [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+
+func TestRun(t *testing.T) {
+	t.Run("success", func(t *testing.T) {
+		tmpFile, err := ioutil.TempFile("", "")
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer tmpFile.Close()
+
+		testInputs := []string{
```
Comment (tk3fftk) [inline buildlog/buildlog_test.go]: We can make this `testInputs` to common.

File: buildlog/buildlog_test.go [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+	t.Run("failure by parsing error", func(t *testing.T) {
+		tmpFile, err := ioutil.TempFile("", "")
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer tmpFile.Close()
+
+		testInputs := []string{
+			`{"t": 1580198209, "m": "test 1", "n": 0, "s": "main"}` + "\n",
+			`{"t": 1580198222, "m": "test 2", "n": 1, "s": "main"` + "\n",
```
Comment (tk3fftk) [inline buildlog/buildlog_test.go]: The input value of this test case looks similar with the success case. 
In my opinion, such test cases make us confusing. Failure case should be more simple if we don't want to test JSON parse.

Comment (kkisic) [reply buildlog/buildlog_test.go]: I fixed it.

File: buildlog/buildlog_test.go (hunk 13 above)
Comment (kkisic) [reply buildlog/buildlog_test.go]: I fixed it.

File: buildlog/buildlog_test.go (hunk 11 above)
Comment (kkisic) [reply buildlog/buildlog_test.go]: I fixed it.

File: buildlog/buildlog.go (hunk 12 above)
Comment (kkisic) [reply buildlog/buildlog.go]: It caused both cases.

Comment (tk3fftk) [reply buildlog/buildlog.go]: thanks!

File: buildlog/buildlog_test.go [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+)
+
+var (
+	testInputs = []string{
+		`{"t": 1581662022394, "m": "test 1", "n": 0, "s": "main"}` + "\n",
+		`{"t": 1581662022395, "m": "test 2", "n": 1, "s": "main"}` + "\n",
+	}
+)
+
+type writeCloser struct {
```
Comment (yuichi10) [inline buildlog/buildlog_test.go]: we can delete this struct and `Close` function

File: buildlog/buildlog_test.go [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+	t.Run("failure by parsing error", func(t *testing.T) {
+		tmpFile, err := ioutil.TempFile("", "")
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer tmpFile.Close()
+
+		testInvalidInputs := []string{
+			`{"t": 1581662022394, "m": "test 1", "n": 0, "s": "main"}` + "\n",
+			`{` + "\n",
```
Comment (yuichi10) [inline buildlog/buildlog_test.go]: Could you check the log package are working after broken line was come?

File: buildlog/buildlog.go [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+
+	for {
+		select {
+		case <-l.ctx.Done():
+			break
+		default:
+			err := l.output(reader)
+			if err != nil {
+				fmt.Printf("failed to run logger: %w\n", err)
+				l.cancel()
```
Comment (yuichi10) [inline buildlog/buildlog.go]: Can we cancel with every error?
I think it stop logging, when failed to parse log.

Comment (yuichi10) [reply buildlog/buildlog.go]: Or can we add the function of continue writing log at second pass?

File: buildlog/buildlog_test.go (hunk 15 above)
Comment (kkisic) [reply buildlog/buildlog_test.go]: I fixed it.

File: buildlog/buildlog_test.go (hunk 16 above)
Comment (kkisic) [reply buildlog/buildlog_test.go]: If broken line is come, logger is stopped, but build is running.

File: buildlog/buildlog.go (hunk 17 above)
Comment (kkisic) [reply buildlog/buildlog.go]: > Can we cancel with every error?
> I think it stop logging, when failed to parse log.

I think it work that way.

> Or can we add the function of continue writing log at second pass?

If broken line is come, logger is stopped, but build is running.
So we have no plan to do so for now.

Comment (yuichi10) [reply buildlog/buildlog.go]: I see. Someday I want these function. Users are hard to understand what happen, if there are no formatted log even the build is running.

**Pull request merged**